### DISCUSSION TITLE
Include headers in cached requests

### DISCUSSION THREAD

**Discussion by 0x524c** [moleculerjs/moleculer#1239] [General]
Hey guys.

I have a project that depends on a header included in the responses of an action. I need to enable a cache for these actions and would like to use Moleculer's native cache feature, but I noticed that Moleculer cache only the response without including the headers.

Is there any option to include the response headers in the cache?

**Comment by icebob** [comment]
You can define response headers inside the action via `ctx.meta.$responseHeaders`. More info: https://moleculer.services/docs/0.14/moleculer-web#Response-type-amp-status-code

**Reply by 0x524c** [reply]
@icebob  Thank you very much. Where is the best place for me to make a donation? (it will be symbolic, but it's a way to recognize the great project you've developed, from one developer to another).

**Reply by icebob** [reply]
You're welcome. 
For donation, you can use [Github Sponsors](https://github.com/sponsors/moleculerjs) or [Open Collective](https://opencollective.com/moleculer). Thanks in advance!

**Comment by 0x524c** [comment]
> You can define response headers inside the action via `ctx.meta.$responseHeaders`. More info: https://moleculer.services/docs/0.14/moleculer-web#Response-type-amp-status-code

Yes, I'm using the meta to add headers as responses (0.14). What I noticed is that when caching is enabled, these custom headers are not saved along with the body.

**Reply by icebob** [reply]
Yes, you are right. Plz open an issue about it, however, I won't able to fix it, just with breaking changes because need to change the whole cache storage mechanism, but I will able to do it in later versions. 

For a workaround, if you split your action to separated action, and the first one is not cached, just calls the second one and put the headers, and the second one is cached and do the logic.

**Reply by 0x524c** [reply]
great, thanks!

I opened a issue: https://github.com/moleculerjs/moleculer/issues/1242

**Reply by intech** [reply]
Or you can use [hook onAfterCall](https://moleculer.services/docs/0.14/moleculer-web#Route-hooks) in the API gateway
